wandb/apis/importers/wandb.py

```diff
@@ -0,0 +1,1879 @@
+        ...
+
+    def program(self) -> Optional[str]:
+        ...
+
+    def host(self) -> Optional[str]:
+        return self._metadata_file().get("host")
+
+    def username(self) -> Optional[str]:
+        ...
+
+    def executable(self) -> Optional[str]:
+        ...
+
+    def gpus_used(self) -> Optional[str]:
+        ...
+
+    def cpus_used(self) -> Optional[int]:  # can we get the model?
+        ...
+
+    def memory_used(self) -> Optional[int]:
+        ...
+
+    def runtime(self) -> Optional[int]:
+        wandb_runtime = self.run.summary.get("_wandb", {}).get("runtime")
+        base_runtime = self.run.summary.get("_runtime")
+
+        if (t := coalesce(wandb_runtime, base_runtime)) is None:
+            return t
+        return int(t)
+
+    def start_time(self) -> Optional[int]:
+        t = dt.fromisoformat(self.run.created_at).timestamp() * 1000
+        return int(t)
+
+    def code_path(self) -> Optional[str]:
+        path = self._metadata_file().get("codePath", "")
+        return f"code/{path}"
+
+    def cli_version(self) -> Optional[str]:
+        return self._config_file().get("_wandb", {}).get("value", {}).get("cli_version")
+
+    def files(self) -> Optional[Iterable[Tuple[str, str]]]:
+        if self._files is not None:
+            yield from self._files
+            return
+
+        files_dir = f"./wandb-importer/{self.run_id()}/files"
+
+        self._files = []
+        for f in self.run.files():
+            f: File
+            # These optimizations are intended to avoid rate limiting when importing many runs in parallel
+            # Don't carry over empty files
+            if f.size == 0:
+                continue
+            # Skip deadlist to avoid overloading S3
+            if "wandb_manifest.json.deadlist" in f.name:
+                continue
+
+            # Download and log file (is this the right wording?)
+            file_and_policy = None
+            result = f.download(files_dir, exist_ok=True, api=self.api)
+            file_and_policy = (result.name, "end")
+            self._files.append(file_and_policy)
+            yield file_and_policy
+
+    def logs(self) -> Optional[Iterable[str]]:
+        if (fname := self._find_in_files("output.log")) is None:
+            return
+
+        with open(fname) as f:
+            yield from f.readlines()
+
+    def _metadata_file(self) -> Dict[str, Any]:
+        if (fname := self._find_in_files("wandb-metadata.json")) is None:
+            return {}
+
+        with open(fname) as f:
+            return json.loads(f.read())
+
+    def _config_file(self) -> Dict[str, Any]:
+        if (fname := self._find_in_files("config.yaml")) is None:
+            return {}
+
+        with open(fname) as f:
+            return yaml.safe_load(f) or {}
+
+    def _get_rows_from_parquet_history_paths(self) -> None:
+        if self._parquet_history_paths is None:
+            self._parquet_history_paths = list(self._get_parquet_history_paths())
```

Comment:
looks like a lot of the new code is untested, see https://app.codecov.io/gh/wandb/wandb/pull/6897?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=wandb (almost 400 missed lines in this file, only 62% coverage). up to you, but...from the future maintenance perspective, I'd either pump up the coverage, or removed the untested code.

Reply:
Yep, sorry.  The "verify" code doesn't run unless the importer is called in a loop.  Added that back and coverage is now ~80%